src/ligero/prover.rs

```diff
@@ -1,3 +1,566 @@
 //! Ligero prover, specified in [Section 4.4][1].
 //!
 //! [1]: https://datatracker.ietf.org/doc/html/draft-google-cfrg-libzk-01#section-4.4
+
+use crate::{
+    Codec,
+    constraints::proof_constraints::{
+        LinearConstraintLhsTerm, LinearConstraints, QuadraticConstraint,
+    },
+    fields::{CodecFieldElement, LagrangePolynomialFieldElement},
+    ligero::{
+        TableauLayout,
+        committer::LigeroCommitment,
+        extend,
+        merkle::{InclusionProof, MerkleTree},
+    },
+    transcript::Transcript,
+};
+use anyhow::anyhow;
+
+/// Prover for Ligero.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct LigeroProver<'a> {
+    tableau_layout: &'a TableauLayout<'a>,
+}
+
+impl<'a> LigeroProver<'a> {
+    /// Construct a new prover from the Ligero parameters.
+    pub fn new(tableau_layout: &'a TableauLayout<'a>) -> Self {
+        Self { tableau_layout }
+    }
+
+    /// Prove that the commitment satisfies the provided constraints. The provided transcript should
+    /// have been used in [`LinearConstraints::from_proof`] (or, equivalently,
+    /// [`SumcheckProver::prove`]).
+    ///
+    /// This is specified in [4.4][1].
+    ///
+    /// [1]: https://datatracker.ietf.org/doc/html/draft-google-cfrg-libzk-01#section-4.4
+    pub fn prove<FE: CodecFieldElement + LagrangePolynomialFieldElement>(
+        &mut self,
+        transcript: &mut Transcript,
+        commitment: LigeroCommitment,
+        tableau: &[Vec<FE>],
+        merkle_tree: &MerkleTree,
+        merkle_tree_nonces: &[[u8; 32]],
+        linear_constraints: &LinearConstraints<FE>,
+        quadratic_constraints: &[QuadraticConstraint],
+    ) -> Result<LigeroProof<FE>, anyhow::Error> {
+        transcript.write_byte_array(commitment.as_bytes())?;
+
+        // The challenge is also "u" in the specification. Generate one challenge element for each
+        // witness and quadratic witness row in the tableau.
+        let challenge = transcript.generate_challenge::<FE>(
+            tableau.len() - self.tableau_layout.first_linear_constraint_row(),
+        )?;
```

Comment:
FYI, this is missing the DEADBEEF message written to the transcript, mentioned here google/longfellow-zk#97.